### PR TITLE
 [Apple] Check for cancellation_date_ms first

### PR DESCRIPTION
**Pull request by superandrew213**
Apple says this about subscription cancellations:
https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/StoreKitGuide/Chapters/Subscriptions.html#//apple_ref/doc/uid/TP40008267-CH7-SW5

If a user cancels the subscription
>  Customers may cancel a subscription in the middle of a subscription period, but the subscription remains paid through the end of the same period.

If Apple customer support cancels the subscription and provides a refund
>To check whether a purchase has been canceled by Apple Customer Support, look for the Cancellation Date field in the receipt. If the field contains a date, regardless of the subscription’s expiration date, the purchase has been canceled. With respect to providing content or service, treat a canceled transaction the same as if no purchase had ever been made.

Based on this I think we should set the `expirationDate` to the cancellation date if a cancellation date is provided by Apple.

This is how a receipt looks like with a cancellation date:
```javascript
    {
        "original_purchase_date_pst": "2018-01-25 11:59:25 America/Los_Angeles",
        "cancellation_date_ms": "1517150504000",
        "cancellation_reason": "0",
        "original_purchase_date_ms": "1516910365000",
        "expires_date_formatted": "2019-01-25 19:59:23 Etc/GMT",
        "is_in_intro_offer_period": "false",
        "purchase_date_ms": "1516910363000",
        "expires_date_formatted_pst": "2019-01-25 11:59:23 America/Los_Angeles",
        "is_trial_period": "false",
        "expires_date": "1548446363000",
        "cancellation_date": "2018-01-28 14:41:44 Etc/GMT",
        "purchase_date": "2018-01-25 19:59:23 Etc/GMT",
        "cancellation_date_pst": "2018-01-28 06:41:44 America/Los_Angeles",
        "purchase_date_pst": "2018-01-25 11:59:23 America/Los_Angeles",
        "original_purchase_date": "2018-01-25 19:59:25 Etc/GMT"
    }
```

@KLVTZ @ronkorving what do you think?

### CODE REVIEW

**Comment by superandrew213** [comment]
Maybe we just need to check `latest_expired_receipt_info` for the cancellation date. Not latest_receipt_info.

**Comment by justinpage** [comment]
@superandrew213 I'll try and reserve some time this week to review. Thanks!

**Comment by justinpage** [comment]
@superandrew213 Wow! I am a year late on this one xD 

Is this still applicable? Can you see if the rules described above have changed?

**Comment by superandrew213** [comment]
@justinpage rules are still the same. Essentially what it means is that if there is `latestExpiredReceiptInfo` data and it contains a `cancellation_date_ms` then the subscription was refunded by Apple. `cancellation_date_ms` is not the date that the user cancelled the subscription auto-renew. It is the date that Apple cancelled the `purchase` and issued the refund.

If there is no `cancellation_date_ms` then the user just cancelled the subscription auto-renew but should still have access until the subscription expires because they were not refunded.

By `cancel`, Apple means cancelled purchase (and refunded subscription payment). It does not refer to cancelling the subscription auto-renew (even though they do this too).

It's all pretty confusing because they did not name things properly. 

Just read this a few times, then it kind of makes sense:

> To check whether a purchase has been canceled by Apple Customer Support, look for the Cancellation Date field in the receipt. If the field contains a date, regardless of the subscription’s expiration date, the purchase has been canceled. With respect to providing content or service, treat a canceled transaction the same as if no purchase had ever been made.

If there is a `cancellation_date_ms` then the subscription was refunded by Apple. We should set this as the `expirationDate` so that user's access is revoked immediately instead of when the subscription expires. Subscription expiration could be in a months time or in a year for yearly subscriptions.

File: lib/apple/index.js [hunk 1]
```diff
@@ -74,19 +74,28 @@ function parseResult(result) {
...
 			purchaseDate = parseTime(result.latest_receipt_info.purchase_date_ms);
 			expirationDate = parseTime(
-				result.latest_receipt_info.expires_date_ms || result.latest_receipt_info.expires_date
+				result.latest_receipt_info.expires_date_ms ||
+				result.latest_receipt_info.expires_date
 			);
 		}
 	}
 
+	// Check for cancellation date
```
Comment (ronkorving) [inline lib/apple/index.js]: Maybe a bit more information here, would be wise? Seeing as the amount of confusion exists around why this is a good idea :)

**Comment by ronkorving** [comment]
Thanks @superandrew213 , time to merge, I think :)
Good things come to those who wait, I guess...

**Comment by superandrew213** [comment]
@ronkorving no worries 👍

**Pull request merged**